### PR TITLE
Implementes filter delay and adds a few keybindings

### PR DESCRIPTION
**Pull request by andrejborstnik**
### Disclaimer
While working on a project I found myself needing a few expansions to the ```awesomplete```.  Since I saw a few other people were asking for similar features I submit this small contribution. I invite you to comment on any feature.
### Changes
* The Up/Down arrows now open the ```ul```. Such a feature is needed if we want complete (and convenient) mouse-less control. E.g. I navigate (with Tab) to an ```input``` with something already written and want to see recommendations. Or I have just clicked Escape, but don't want to change the text to get the ```ul``` back.
* The filtered list remembers where we left off (we just don't set ```index = -1``` on close). It seems only natural to me to continue highlighting where we left off last time, if the filtered list has not changed.
* Added the Tab and Shift-Tab keybindings. For me tabbing while inside the ```ul``` is natural. Regarding the concern about navigating inputs with Tab: Tab does not open the ```ul```, so by selecting a value (enter), or by closing the ```ul``` (Escape) we are free to navigate away. I also considered #16832, but I never use Tab to select like that - it only seems natural if the completed text is shadowed in the input field, or there is only one choice (as the command line). But if your experience is different it is an easy fix.
* Added the option to block filtering for a certain amount of time (after each filter), which is necessary for large lists (```100k+```). The default ```0.5s``` seems feasible to me, but feel free to change it to 0 and it will not affect the operation of ```awesomplete```. Also solves #16887.
### Tests
I updated the tests to reflect the changes. I did not (yet) write tests for delayed filtering, but I have tested it. Existing tests pass for ```delay = 0```, but would need to be adapted for other delays, as not every call of the (old) evaluate function is (fully) executed.

### CODE REVIEW

**Comment by LeaVerou** [comment]
Thank you so much for this contribution @andrejborstnik, this looks great! What do the others think? Should we merge?

Small note: In the future, it would be much easier for us if you opened separate PRs for separate improvements. It would also mean your PRs get merged faster, because reviewing them is easier.

**Comment by LeaVerou** [comment]
Since this modifies the keyboard UX quite a bit, I would love to try it out and see how it feels (I'm a heavy keyboard user too). Is there any online demo? I tried https://andrejborstnik.github.io/awesomplete/ but it doesn't seem to have the improvements (e.g. pressing down arrow on an input does nothing).

File: awesomplete.js [hunk 1]
```diff
@@ -236,6 +248,25 @@ _.prototype = {
...
+					setTimeout(function () {
+							return this.evaluate();
+						}.bind(this),
+						this.filterDelay);
+					this.filterDelayed = true;
+				}
+				return;
+			}
+			this.lastFilterTime = time;
+			this.filterDelayed = false;
```
Comment (LeaVerou) [inline awesomplete.js]: Do we really need a separate flag for this? Wouldn't checking `this.filterDelay > 0)` suffice?

**Comment by andrejborstnik** [comment]
@LeaVerou I will keep multiple PRs in mind next time :). 

The demo on  https://andrejborstnik.github.io/awesomplete/ does have all the improvements, but does not showcase delay. If you want I can write up a demo for that. You can notice the default delay though if you compare it to https://leaverou.github.io/awesomplete/ (delete chars fast and see a delay on my page).

You are not seeing anything on arrow down, because there is nothing in the input - the arrow down only shows recommendations if other conditions are met (e.g. the number of chars written). I programmed it this way, so that the changes to program logic are minimal.

**Comment by vlazar** [comment]
First of all I would 👍 splitting this into separate PRs but we can discuss proposed features first.

Changes with keyboard UX may not be what all users of Awesomplete want. I think we should consider how difficult it would be to add proposed changes with tab/shift-tab without modification of Awesomplete itself versus how difficult it would be to get current behavior with tab/shift-tab if proposed changes are merged in.

Not sure about remembering highlighted item. But again, we need to evaluate it if it's something most users of Awesomplete would expect.

Delayed filtering looks interesting, although personally I would prefer a more generic solution, e.g. adding support for any async suggestion lists. We have a corresponding list of issues https://github.com/LeaVerou/awesomplete/milestone/4

**Comment by andrejborstnik** [comment]
No problem, when we converge to a features list I will create seperate pull requests.

The tab/shift tab could be enabled with an additional option parameter if this feature is not desired by all users - we can than optionally create a poll and see how many people use it. I know I would never expect tab to shift fields while inside the dropdown. But as I stated this was created for my project and is subject to other opinions.

This could again be an option parameter.

The delay was designed for two reasons: to stop unnecessary computation and to improve the UX by reducing browser lag with big lists. 

Async could (if the call only returned the matched values) solve the second problem. But than I only use awesomplete to display, not to autocomplete. 

And async would still not solve the first problem: each time* the input changes we would compute the suggestions. If this is done on the frontend the user might see browser delays and if we do it on the backend we put a lot of unnecessary strain on the server. 

By blocking requests for a period of time we solve both problems at the same time. This is already made optional, as we retain the same functionality by setting delay to 0 (could be default).

**Pull request closed without merging**